src/account_manager.rs

```diff
@@ -110,59 +143,95 @@ impl AccountManager {
             None,
         )?;
         crate::init_stronghold(&self.storage_path, stronghold);
-        self.start_background_sync();
+
+        if self.accounts.read().await.is_empty() {
+            self.accounts = Self::load_accounts(&self.storage_path).await?;
+            let _ = self.start_monitoring().await;
+        }
+
         Ok(())
     }
 
     /// Starts the polling mechanism.
-    fn start_polling(&self, is_monitoring_disabled: bool) -> thread::JoinHandle<()> {
+    fn start_polling(&self, is_monitoring_disabled: bool, mut stop: BroadcastReceiver<()>) {
         let storage_path = self.storage_path.clone();
-        let interval = self.polling_interval;
+        let accounts = self.accounts.clone();
+
+        let interval = AsyncDuration::from_millis(self.polling_interval.as_millis().try_into().unwrap());
+
         thread::spawn(move || {
-            loop {
-                let storage_path_ = storage_path.clone();
-                crate::block_on(async move {
-                    if let Err(panic) = AssertUnwindSafe(poll(storage_path_, is_monitoring_disabled))
-                        .catch_unwind()
-                        .await
-                    {
-                        let msg = if let Some(message) = panic.downcast_ref::<String>() {
-                            format!("Internal error: {}", message)
-                        } else if let Some(message) = panic.downcast_ref::<&str>() {
-                            format!("Internal error: {}", message)
-                        } else {
-                            "Internal error".to_string()
-                        };
-                        let _error = crate::WalletError::UnknownError(msg);
-                        // when the error is dropped, the on_error event will be triggered
+            crate::enter(|| {
+                tokio::spawn(async move {
+                    loop {
+                        tokio::select! {
+                            _ = async {
+                                let storage_path_ = storage_path.clone();
+                                let accounts_ = accounts.clone();
+
+                                if let Err(panic) = AssertUnwindSafe(poll(accounts_.clone(), storage_path_, is_monitoring_disabled))
+                                    .catch_unwind()
+                                    .await {
+                                    let msg = if let Some(message) = panic.downcast_ref::<String>() {
```

Comment:
Using `Cow<str>` to allow for both `String` and `str` without needing dynamic dispatching and downcasting might be a good idea (the code seems to imply that the panic message being neither a `&'static str` nor a `String` is a type system hole: `Cow` will let you patch it)

Reply:
I'll check again if this code really needs a catch_unwind; I think with this refactor it won't be needed anymore.

But I do have this check on the Node.js binding and unfortunately I can't remove it there, so I guess i'll write something to improve it.